core/core_test.go

```diff
@@ -85,6 +85,70 @@ func TestTranscode(t *testing.T) {
 	// TODO check transcode loop expiry, storage, sig construction, etc
 }
 
+func StubTranscodeConfig() transcodeConfig {
+	return transcodeConfig{
+		OS:      nil,
+		LocalOS: nil,
+	}
+}
+
+func TestTranscodeSeg_OffchainMode_ReturnsEmptySignature(t *testing.T) {
+	//Set up the node
+	drivers.NodeStorage = drivers.NewMemoryDriver("")
+	db, err := common.InitDB("file:TestTranscode?mode=memory&cache=shared")
+	require := require.New(t)
+	require.Nil(err)
+	defer db.Close()
+
+	tmp, err := ioutil.TempDir("", "")
+	require.Nil(err)
+	n, err := NewLivepeerNode(nil, tmp, db)
+	require.Nil(err)
+
+	// mock transcoder
+	n.Transcoder = NewLocalTranscoder(tmp)
+	ffmpeg.InitFFmpeg()
+	defer os.RemoveAll(tmp)
+
+	// mock transcodeSeg
+	transcodeConfig := StubTranscodeConfig()
+	segment := StubSegment()
+	segmentMetadata := &SegTranscodingMetadata{Profiles: videoProfiles}
+	transcodeResult := n.transcodeSeg(transcodeConfig, segment, segmentMetadata)
+
+	assert := assert.New(t)
+	assert.Nil(transcodeResult.Sig)
+}
+
+func TestTranscodeSeg_OnchainMode_ReturnsSignature(t *testing.T) {
```

Comment:
To me, this seems like a good example of where the style of breaking cases into single, specific tests does not facilitate overall understanding of the cases under test. Too much boilerplate, and the high-level understanding of the test coverage is inhibited by the loss of locality.

How does this look in terms of a halfway point between "whole function testing" and "single case testing"? This also incorporates most of the feedback from elsewhere in this review.

```diff
diff --git a/core/core_test.go b/core/core_test.go
index 7cc06889..01ca7d0e 100644
--- a/core/core_test.go
+++ b/core/core_test.go
@@ -82,7 +82,45 @@ func TestTranscode(t *testing.T) {
                t.Error("Unexpected transcode result ", len(tr.Data[1]))
        }
 
-       // TODO check transcode loop expiry, storage, sig construction, etc
+       // TODO check transcode loop expiry, storage, etc
+}
+
+func TestTranscodeSig(t *testing.T) {
+       tmp, _ := ioutil.TempDir("", "")
+       defer os.RemoveAll(tmp)
+
+       p := []ffmpeg.VideoProfile{ffmpeg.P720p60fps16x9, ffmpeg.P144p30fps16x9}
+       n, _ := NewLivepeerNode(nil, tmp, nil)
+       n.Transcoder = &StubTranscoder{Profiles: p}
+       conf := transcodeConfig{LocalOS: (drivers.NewMemoryDriver("")).NewSession("")}
+       md := &SegTranscodingMetadata{Profiles: p}
+       seg := StubSegment()
+
+       assert := assert.New(t)
+       require := require.New(t)
+
+       // Test offchain mode
+       require.Nil(n.Eth) // sanity check the offchain precondition of a nil eth
+       res := n.transcodeSeg(conf, seg, md)
+       assert.Nil(res.Err)
+       assert.Nil(res.Sig)
+       // sanity check results
+       resBytes, _ := n.Transcoder.Transcode("", p)
+       assert.Equal(resBytes, res.Data)
+
+       // Test onchain mode
+       n.Eth = &eth.StubClient{}
+       res = n.transcodeSeg(conf, seg, md)
+       assert.Nil(res.Err)
+       assert.NotNil(res.Sig)
+       // check sig
+       resHashes := make([][]byte, len(p))
+       for i, v := range resBytes {
+               resHashes[i] = ethCrypto.Keccak256(v)
+       }
+       resHash := ethCrypto.Keccak256(resHashes...)
+       assert.Equal(resHash, res.Sig)
 }
 
 func TestTranscodeLoop_GivenNoSegmentsPastTimeout_CleansSegmentChan(t *testing.T) {

diff --git a/eth/stubclient.go b/eth/stubclient.go
index 18cba6de..5e6f2ad2 100644
--- a/eth/stubclient.go
+++ b/eth/stubclient.go
@@ -260,7 +260,7 @@ func (c *StubClient) CheckTx(tx *types.Transaction) error {
 func (c *StubClient) ReplaceTransaction(tx *types.Transaction, method string, gasPrice *big.Int) (*types.Transaction, error) {
        return nil, nil
 }
-func (c *StubClient) Sign(msg []byte) ([]byte, error)   { return nil, nil }
+func (c *StubClient) Sign(msg []byte) ([]byte, error)   { return msg, nil }
```

Reply:
Thank you for the example Josh! Really helpful.